Userland/Libraries/LibDeviceTree/DeviceTree.h

```diff
@@ -0,0 +1,191 @@
+    HashMap<StringView, DeviceTreeProperty>& properties() { return m_properties; }
+    DeviceTreeNodeView* parent() { return m_parent; }
+
+private:
+    DeviceTreeNodeView* m_parent;
+    HashMap<StringView, DeviceTreeNodeView> m_children;
+    HashMap<StringView, DeviceTreeProperty> m_properties;
+};
+
+class DeviceTree : public DeviceTreeNodeView {
+public:
+    static ErrorOr<NonnullOwnPtr<DeviceTree>> parse(ReadonlyBytes);
+
+    DeviceTreeNodeView const* resolve_node(StringView path) const
+    {
+        // FIXME: May children of aliases be referenced?
+        if (auto alias_list = get_child("aliases"sv); alias_list.has_value()) {
+            if (auto alias = alias_list->get_property(path); alias.has_value()) {
+                path = alias.value().as_string();
+            }
+        }
+
+        DeviceTreeNodeView const* node = this;
+        path.for_each_split_view('/', SplitBehavior::Nothing, [&](auto const& part) {
+            if (auto child = node->get_child(part); child.has_value()) {
+                node = &child.value();
+            } else {
+                node = nullptr;
+                return IterationDecision::Break;
+            }
+            return IterationDecision::Continue;
+        });
+
+        return node;
+    }
+
+    Optional<DeviceTreeProperty> resolve_property(StringView path) const
+    {
+        auto property_name = path.find_last_split_view('/');
+        auto node_path = path.substring_view(0, path.length() - property_name.length() - 1);
+        auto const* node = resolve_node(node_path);
+        if (!node)
+            return {};
+        return node->get_property(property_name);
+    }
+
+    // FIXME: Add a helper to iterate over each descendant fulfilling some properties
+    //        Like each node with a "compatible" property containing "pci" or "usb",
+    //        bonus points if it could automatically recurse in the tree under some conditions,
+    //        like "simple-bus" or "pci-bridge" nodes
+
+    DeviceTreeNodeView const* phandle(u32 phandle) const
+    {
+        if (phandle >= m_phandles.size())
+            return nullptr;
+        return m_phandles[phandle];
+    }
+
+    ReadonlyBytes flattened_device_tree() const { return m_flattened_device_tree; }
+
+private:
+    DeviceTree(ReadonlyBytes flattened_device_tree)
+        : DeviceTreeNodeView(nullptr)
+        , m_flattened_device_tree(flattened_device_tree)
+    {
+    }
+
+    ErrorOr<void> set_phandle(u32 phandle, DeviceTreeNodeView* node)
+    {
+        if (m_phandles.size() > phandle && m_phandles[phandle] != nullptr)
+            return Error::from_string_view_or_print_error_and_return_errno("Duplicate phandle entry in DeviceTree"sv, EINVAL);
+        if (m_phandles.size() <= phandle)
+            TRY(m_phandles.try_resize(phandle + 1));
+        m_phandles[phandle] = node;
+        return {};
+    }
+
+    ReadonlyBytes m_flattened_device_tree;
+    Vector<DeviceTreeNodeView*> m_phandles;
```

Comment:
Can't the phandle list contain references?

Reply:
As I dont trust firmware to hand out phandles in order, I wanted to allow empty entries
And while they seem to be handed out in order starting with 1 - so slot 0 is always a `nullptr` anyway - I don't think the spec really enforces that

Reply:
Yeah, then keep them pointers.